### PR TITLE
Fix #17965 datetime64 comparison

### LINKED ISSUE #17965: COMPAT: datetime.date comparisons with datetime64[ns]
This should probably work
```
In [16]: df = pd.DataFrame({'A':[pd.Timestamp('20130101'), np.nan]})

In [17]: df
Out[17]: 
           A
0 2013-01-01
1        NaT

In [18]: pd.Timestamp('20130101').to_pydatetime()>=df.A
Out[18]: 
0     True
1    False
Name: A, dtype: bool

In [20]: pd.Timestamp('20130101').to_pydatetime().date()>=df.A
ValueError: cannot set a Timestamp with a non-timestamp
```

### PR DESCRIPTION
**Pull request by watercrossing**
- [x] closes #17965 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_partial_slicing.py [hunk 1]
```diff
@@ -330,3 +330,9 @@ def test_loc_datetime_length_one(self):
 
         result = df.loc['2016-10-01T00:00:00':]
         tm.assert_frame_equal(result, df)
+
+    def test_comparison_to_datetime64(self):
+        # GH issue #17965
+        df = DataFrame({'dates': date_range('2000-01-01', '2000-01-05')})
+        date = df['dates'].unique()[0]
```
Comment (TomAugspurger) [inline pandas/tests/indexes/datetimes/test_partial_slicing.py]: Is this the right test? `date` should be a `datetime.date`, and I think this is a `datetime64[ns]`.

**Comment by TomAugspurger** [comment]
Python is inconsistent here:

```python
python                                                                                                                                                                                                                               Thu Nov  9 05:53:21 2017
Python 3.6.1 (default, Apr  4 2017, 09:40:21)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import datetime
>>> datetime.datetime(2017, 1, 1) == datetime.date(2017, 1, 1)
False
>>> datetime.datetime(2017, 1, 1) >= datetime.date(2017, 1, 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't compare datetime.datetime to datetime.date
```

Strange.... We should be a bit clearer on our policy if we're differing from Python. Are we saying that `datetime.datetimes` compare with a `datetime.date` by promoting the `date` to a datetime at midnight?

File: pandas/tests/indexes/datetimes/test_partial_slicing.py (hunk 1 above)
Comment (watercrossing) [reply pandas/tests/indexes/datetimes/test_partial_slicing.py]: date is `numpy.datetime64('2000-01-01T00:00:00.000000000')`, which is consistent with the output of `pd.date_range().values`, which are `datetime64[ns]` too.  Do you suggest this should be an array of `datetime.date` objects?

Comment (TomAugspurger) [reply pandas/tests/indexes/datetimes/test_partial_slicing.py]: #17965 is about comparing with a single `datetime.date` object:

```python
In [20]: pd.Timestamp('20130101').to_pydatetime().date()>=df.A
ValueError: cannot set a Timestamp with a non-timestamp
```

**Comment by watercrossing** [comment]
I don't think your comparison is correct - semantically the test is comparing two `date`s, not `datetimes`:
`datetime.date(2013,11,4) == datetime.date(2013,11,4)` is `True` - so we are not deviating from Python?

File: doc/source/whatsnew/v0.21.1.txt [hunk 2]
```diff
@@ -61,6 +61,7 @@ Bug Fixes
 - Bug in :class:`DatetimeIndex` subtracting datetimelike from DatetimeIndex could fail to overflow (:issue:`18020`)
 - Bug in ``pd.Series.rolling.skew()`` and ``rolling.kurt()`` with all equal values has floating issue (:issue:`18044`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
+- Bug in ``_libs.index.convert_scalar()`` when comparing datetimes to np.datetime64 (:issue:`17965`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: make a user useful note

File: pandas/_libs/index.pyx [hunk 3]
```diff
@@ -549,6 +549,8 @@ cpdef convert_scalar(ndarray arr, object value):
     if arr.descr.type_num == NPY_DATETIME:
         if isinstance(value, np.ndarray):
             pass
+        if isinstance(value, np.datetime64):
+            return Timestamp(value).value
         elif isinstance(value, datetime):
```
Comment (jreback) [inline pandas/_libs/index.pyx]: compbine np.datetime64 and datetime into the same clause

File: pandas/tests/indexes/datetimes/test_partial_slicing.py (hunk 1 above)
Comment (jreback) [reply pandas/tests/indexes/datetimes/test_partial_slicing.py]: paramaterize this test with ``date, datetime, np.datetime64, and Timestamp``.
check for < and ==

File: pandas/tests/indexes/datetimes/test_partial_slicing.py [hunk 4]
```diff
@@ -330,3 +330,9 @@ def test_loc_datetime_length_one(self):
...
+        tm.assert_frame_equal(df.loc[df['dates'] == date], df.iloc[:1])
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_partial_slicing.py]: use 
```
result =
expected = 
tm.assert_frame_equal(result, expected)
```

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/datetimes/test_partial_slicing.py [hunk 5]
```diff
@@ -330,3 +330,9 @@ def test_loc_datetime_length_one(self):
 
         result = df.loc['2016-10-01T00:00:00':]
         tm.assert_frame_equal(result, df)
+
+    def test_comparison_to_datetime64(self):
+        # GH issue #17965
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_partial_slicing.py]: use the frame from the original issue (which uses a NaT) as well.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/datetimes/test_partial_slicing.py (hunk 1 above)
Comment (watercrossing) [reply pandas/tests/indexes/datetimes/test_partial_slicing.py]: True! I was not actually fixing the Issue, but rather this comment: https://github.com/pandas-dev/pandas/issues/17965#issuecomment-342821038 - but it actually applies to both.

**Comment by jreback** [comment]
@watercrossing the comparison is a non-converted ``datetime.date`` vs a ``Timestamp``, which is what boxed versions of ``datetime64[ns]`` are. pls use the original issues exactly. 

testing against a column of ``datetime.date`` is possible as well, though this is more tricky to construct.

**Comment by jbrockmendel** [comment]
Reiterating suggestion to coerce datetime.date to Period with freq=D

**Comment by jreback** [comment]
> Reiterating suggestion to coerce datetime.date to Period with freq=D

not in this PR maybe at some point

File: doc/source/whatsnew/v0.21.1.txt [hunk 6]
```diff
@@ -61,6 +61,7 @@ Bug Fixes
 - Bug in :class:`DatetimeIndex` subtracting datetimelike from DatetimeIndex could fail to overflow (:issue:`18020`)
 - Bug in ``pd.Series.rolling.skew()`` and ``rolling.kurt()`` with all equal values has floating issue (:issue:`18044`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
+- Bug in ``_libs.index.convert_scalar()``: Comparison between ``datetime64[ns]`` series and ``datetimelike`` (:issue:`17965`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: don't show internal functions here, rather the end-user effect, something like

``a boolean comparison with a datetime.date and a datetime64[ns] dtype Series``

File: pandas/tests/indexes/datetimes/test_partial_slicing.py [hunk 7]
```diff
@@ -330,3 +332,23 @@ def test_loc_datetime_length_one(self):
 
         result = df.loc['2016-10-01T00:00:00':]
         tm.assert_frame_equal(result, df)
+
+    comparisonTS = Timestamp('20130101')
+
+    @pytest.mark.parametrize('datetimelike,opExp', product(
+        [comparisonTS, comparisonTS.to_pydatetime(),
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_partial_slicing.py]: directly construct these, e.g. ``Timestamp('20130101'), datetime.datetime(2013, 1, 1)`` etc

File: pandas/tests/indexes/datetimes/test_partial_slicing.py [hunk 8]
```diff
@@ -330,3 +332,23 @@ def test_loc_datetime_length_one(self):
...
+         comparisonTS.to_pydatetime().date(), comparisonTS.to_datetime64()],
+        [(op.lt, [True, False, False, False]),
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_partial_slicing.py]: don't pass ``opExp``, rather pass ``op`` (e.g. the op.lt) and the ``expected`` (the boolean result)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/datetimes/test_partial_slicing.py (hunk 8 above)
Comment (watercrossing) [reply pandas/tests/indexes/datetimes/test_partial_slicing.py]: Can do, but i am not sure something like:
```
[(x, y[0], y[1]) for x, y in product( 
    [pd.Timestamp('20130101'), datetime.datetime(2013, 1, 1),
     datetime.date(2013, 1, 1), np.datetime64("2013-01-01", 'ns')],
    [(op.lt, [True, False, False, False])])]
```
is more readable.

Comment (jreback) [reply pandas/tests/indexes/datetimes/test_partial_slicing.py]: sorry didn't see the product. instead use multiple paramatrize to achieve product.

eg..
```
@pytest.mark.parametrize(.....)
@pytest.mark.parametrize('datetimelike'
```

**Comment by jreback** [comment]
rebase as well

File: doc/source/whatsnew/v0.21.1.txt [hunk 9]
```diff
@@ -62,6 +62,7 @@ Bug Fixes
 - Bug in ``pd.Series.rolling.skew()`` and ``rolling.kurt()`` with all equal values has floating issue (:issue:`18044`)
 - Bug in ``pd.DataFrameGroupBy.count()`` when counting over a datetimelike column (:issue:`13393`)
 - Bug in ``pd.concat`` when empty and non-empty DataFrames or Series are concatenated (:issue:`18178` :issue:`18187`)
+- Bug in a boolean comparison with a ``datetime.datetime``, ``datetime.date``, ``np.datetime64`` and a ``datetime64[ns]`` dtype Series (:issue:`17965`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.1.txt]: just list ``datetime.datetime`` (as that's the actual bug)

**Review by jreback — APPROVED** [review]
small doc-change. ping on green.

File: doc/source/whatsnew/v0.21.1.txt (hunk 9 above)
Comment (jreback) [reply doc/source/whatsnew/v0.21.1.txt]: move to the indexing section

Comment (watercrossing) [reply doc/source/whatsnew/v0.21.1.txt]: @jreback Done.

**Comment by jreback** [comment]
ok ping on green.

**Comment by watercrossing** [comment]
@jreback: I have changed the whatsnew and rebased.

**Comment by TomAugspurger** [comment]
@watercrossing let us know if / when you see all the CI workers finish and we'll merge.

**Comment by jreback** [comment]
> @watercrossing let us know if / when you see all the CI workers finish and we'll merge.

@TomAugspurger optimist!

**Comment by watercrossing** [comment]
@TomAugspurger / @jreback: Ping!

**Comment by jreback** [comment]
thanks @watercrossing

**Pull request merged**